core/src/main/java/io/questdb/griffin/engine/groupby/DistinctSymbolRecordCursorFactory.java

```diff
@@ -0,0 +1,180 @@
+        this.metadata = metadata;
+        this.tableName = tableName;
+        this.columnIndex = columnIndex;
+        this.tableVersion = tableVersion;
+        this.cursor = new DistinctSymbolRecordCursor();
+    }
+
+    @Override
+    public void close() {
+    }
+
+    @Override
+    public RecordCursor getCursor(SqlExecutionContext executionContext) {
+        TableReader reader = engine.getReader(executionContext.getCairoSecurityContext(), tableName, tableVersion);
+        CharSequence columnName = reader.getMetadata().getColumnName(columnIndex);
+        int columnIndex = reader.getMetadata().getColumnIndex(columnName);
+        cursor.of(reader, columnIndex);
+        return cursor;
+    }
+
+    @Override
+    public RecordMetadata getMetadata() {
+        return metadata;
+    }
+
+    @Override
+    public boolean isRandomAccessCursor() {
+        return true;
+    }
+
+    private static class DistinctSymbolRecordCursor implements RecordCursor {
+        private DistinctSymbolRecord recordA = new DistinctSymbolRecord();
+        private DistinctSymbolRecord recordB = new DistinctSymbolRecord();
+        private TableReader reader;
+        private SymbolMapReader symbolMapReader;
+        private int numberOfSymbols;
+
+        DistinctSymbolRecordCursor() {
+        }
+
+        @Override
+        public void close() {
+            Misc.free(reader);
+        }
+
+        @Override
+        public Record getRecord() {
+            return recordA;
+        }
+
+        @Override
+        public SymbolTable getSymbolTable(int columnIndex) {
+            return symbolMapReader;
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (recordA.getRecordIndex() + 1 < numberOfSymbols) {
+                recordA.incrementRecordIndex();
+                return true;
+            }
+            return false;
+        }
+
+        @Override
+        public Record getRecordB() {
+            return recordB;
+        }
+
+        @Override
+        public void recordAt(Record record, long atRowId) {
+            ((DistinctSymbolRecord) record).recordIndex = (int) atRowId;
+        }
+
+        @Override
+        public void toTop() {
+            recordA.recordIndex = -1;
+        }
+
+        public void of(TableReader reader, int columnIndex) {
+            this.symbolMapReader = reader.getSymbolMapReader(columnIndex);
+            this.numberOfSymbols = symbolMapReader.size();
+            this.reader = reader;
+            this.recordA.of(reader, columnIndex);
+            this.recordB.of(reader, columnIndex);
+        }
+
+        @Override
+        public long size() {
+            return numberOfSymbols;
+        }
+    }
+
+    public static class DistinctSymbolRecord implements Record {
+        private int columnIndex;
+        private int recordIndex = -1;
+        private TableReader reader;
+
+        @Override
+        public CharSequence getSym(int col) {
+            if (reader != null) {
```

Comment:
reader null check should not be necessary when Record is setup correctly. If reader is every to become null this should be reported where cursor is requested (don't think null is possible at all in this case)